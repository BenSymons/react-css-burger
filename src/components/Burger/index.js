import React from 'react';
import PropTypes from 'prop-types';

import css from './burger.css';

const Burger = props => {
  const root = document.documentElement;
  const {
    active,
    onClick,
    burger,
    color,
    hoverOpacity,
    scale,
    marginTop,
    marginLeft,
    style,
  } = props;

  root.style.setProperty('--color', color);
  root.style.setProperty('--opacity-hover', hoverOpacity);
  root.style.setProperty('--scale', scale);
  root.style.setProperty('--top-margin', marginTop);
  root.style.setProperty('--left-margin', marginLeft);

  const classes = {
    burger: `${css['hamburger']} ${css[`hamburger--${burger}`]}`,
    box: `${css['hamburger-box']}`,
    inner: `${css['hamburger-inner']}`,
    active: active ? `${css['is-active']}` : '',
  };

  return (
    <button
      onClick={onClick}
      className={`${classes.burger} ${classes.active}`}
      type="button"
      style={style}
    >
      <span className={classes.box}>
        <span className={classes.inner} />
      </span>
    </button>
  );
};

Burger.propTypes = {
  active: PropTypes.bool,
  onClick: PropTypes.func,
  burger: PropTypes.oneOf([
    'arrow',
    'arrowalt',
    'arrowturn',
    'boring',
    'collapse',
    'elastic',
    'emphatic',
    'minus',
    'slider',
    'spin',
    'spring',
    'squeeze',
    'stand',
    'vortex',
    '3dx',
    '3dy',
    '3dxy',
  ]),
  color: PropTypes.string,
  hoverOpacity: PropTypes.number,
  scale: PropTypes.number,
  marginTop: PropTypes.string,
  marginLeft: PropTypes.string,
  style: PropTypes.object,
};

Burger.defaultProps = {
  burger: 'boring',
  color: '#000',
  hoverOpacity: 1,
  scale: 1,
  marginTop: '25px',
  marginLeft: '25px',
};

export default Burger;
