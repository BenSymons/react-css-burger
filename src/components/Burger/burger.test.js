import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import Burger from './';

Enzyme.configure({ adapter: new Adapter() });

describe('<Burger />', () => {
  const wrapper = shallow(<Burger />);
  const wrapperWithProps = shallow(
    <Burger
      onClick={() => console.log('clicked')}
      burger="arrow"
      color="#ada"
      hoverOpacity={0.7}
      scale={0.7}
      marginTop="50px"
      marginLeft="50px"
      style={{ display: 'inline-block' }}
      type="not a button"
      badProp="this is a bad prop"
    />,
  );

  const expectedProps = ['onClick', 'children', 'style', 'type', 'className'];

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render a <button>', () => {
    expect(wrapper.find('button').exists()).toBe(true);
  });

  it('should contain two <span> child elements', () => {
    expect(wrapper.find('span')).toHaveLength(2);
  });

  it('should pass the onClick prop to the <button> element', () => {
    expect(wrapperWithProps.find('button').props().onClick).toBeTruthy();
  });

  it('should pass the style prop to the <button> element', () => {
    expect(wrapperWithProps.find('button').props().style).toBeTruthy();
  });

  it('should NOT reassign the type prop to the <button> element even if passed', () => {
    expect(wrapperWithProps.find('button').props().type).toEqual('button');
  });

  it('should NOT pass any props to the <button> element other than the expected props', () => {
    expect(
      Object.keys(wrapperWithProps.find('button').props()).every(prop =>
        expectedProps.includes(prop),
      ),
    ).toBeTruthy();
  });
});
