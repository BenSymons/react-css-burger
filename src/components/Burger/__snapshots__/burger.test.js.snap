// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Burger /> should render correctly 1`] = `
<button
  className="undefined undefined "
  type="button"
>
  <span
    className="undefined"
  >
    <span
      className="undefined"
    />
  </span>
</button>
`;
